From 82726e9a18658b44925e515b6c1e5dd0163b7574 Mon Sep 17 00:00:00 2001
From: Dong-Yuan Chen <dong-yuan.chen@intel.com>
Date: Sat, 17 Aug 2013 22:19:09 -0700
Subject: Dalvik: PCG: Improve code generation for returns

BZ: 130538

Avoid multiple loads of the self pointer, and make sure to adhere to
PCG conventions for moving values into hard registers prior to an
outgoing JSR.

Category: device-enablement
Domain: AOSP-Dalvik-Compiler-CG
Origin: internal
Upstream-Candidate: no, needs rework

Change-Id: Ia0285e92cf3fdefbe423b40a82e5eed7a4a11361
Orig-MCG-Change-Id: Ibf7d26a08bd43491f9574c6e4090d03d9c2a6d20
Signed-off-by: Kreitzer, David L <david.l.kreitzer@intel.com>
Reviewed-on: http://android.intel.com:8080/125304
Reviewed-by: Lupusoru, Razvan A <razvan.a.lupusoru@intel.com>
Reviewed-by: Hartley, Timothy D <timothy.d.hartley@intel.com>
Reviewed-by: Beyler, Jean Christophe <jean.christophe.beyler@intel.com>
Reviewed-by: Dittert, Eric <eric.dittert@intel.com>
Reviewed-by: lab_aqa <lab_aqa@intel.com>
Reviewed-by: Semukhina, Elena V <elena.v.semukhina@intel.com>
Reviewed-by: Chen, Dong-Yuan <dong-yuan.chen@intel.com>
Tested-by: Chen, Dong-Yuan <dong-yuan.chen@intel.com>
Reviewed-by: cactus <cactus@intel.com>
Tested-by: cactus <cactus@intel.com>
Reviewed-on: http://android.intel.com:8080/125995
Signed-off-by: Serguei Katkov <serguei.i.katkov@intel.com>
---
 vm/compiler/codegen/x86/pcg/LowerCall.cpp |    8 ++++++--
 1 files changed, 6 insertions(+), 2 deletions(-)

diff --git a/vm/compiler/codegen/x86/pcg/LowerCall.cpp b/vm/compiler/codegen/x86/pcg/LowerCall.cpp
index 99258c8..ec4fe6a 100644
--- a/vm/compiler/codegen/x86/pcg/LowerCall.cpp
+++ b/vm/compiler/codegen/x86/pcg/LowerCall.cpp
@@ -301,15 +301,19 @@ void dvmCompilerPcgTranslateReturn (CompilationUnitPCG *cUnit, MIR *mir, bool is
     CGInst saveAreaIl = CGCreateNewInst ("mov", "r", cUnit->getVMPtr ());
     saveAreaIl = CGCreateNewInst ("add", "ri", saveAreaIl,
             -sizeof (StackSaveArea));
-    CGSetRreg (saveAreaIl, "edx");
 
     CGInst prevFrameIl =
         dvmCompilerPcgCreateSimpleLoad (cUnit->getVMPtr (),
                 OFFSETOF_MEMBER (StackSaveArea,prevFrame) -
                 sizeof (StackSaveArea));
+
+    saveAreaIl = CGCreateNewInst ("mov", "r", saveAreaIl);
+    CGSetRreg (saveAreaIl, "edx");
+
+    prevFrameIl = CGCreateNewInst ("mov", "r", prevFrameIl);
     CGSetRreg (prevFrameIl, "edi");
 
-    selfIl = dvmCompilerPcgGetSelfPointer (cUnit);
+    selfIl = CGCreateNewInst ("mov", "r", selfIl);
     CGSetRreg (selfIl, "ecx");
 
     CGInst parmEbp = dvmCompilerPcgGenerateFramePtrMov (cUnit);
-- 
1.7.4.1

